tests/gtest/avifrgbtest.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 Google LLC
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include <tuple>
+
+#include "avif/internal.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+using ::testing::Combine;
+using ::testing::Values;
+
+namespace libavif {
+namespace {
+
+class SetGetRGBATest
+    : public testing::TestWithParam<std::tuple<
+          /*rgb_depth=*/int, avifRGBFormat, /*is_float=*/bool>> {};
+
+TEST_P(SetGetRGBATest, SetGetTest) {
+  const int rgb_depth = std::get<0>(GetParam());
+  const avifRGBFormat rgb_format = std::get<1>(GetParam());
+  const bool is_float = std::get<2>(GetParam());
+
+  // Unused yuv image, simply needed to initialize the rgb image.
+  testutil::AvifImagePtr yuv(
+      avifImageCreate(/*width=*/13, /*height=*/17, 8, AVIF_PIXEL_FORMAT_YUV444),
+      avifImageDestroy);
+
+  testutil::AvifRgbImage rgb(yuv.get(), rgb_depth, rgb_format);
+  rgb.isFloat = is_float;
+
+  avifRGBColorSpaceInfo color_space;
+  ASSERT_TRUE(avifGetRGBColorSpaceInfo(&rgb, &color_space));
+
+  float epsilon = 1.0f / color_space.maxChannelF;
+  if (rgb_format == AVIF_RGB_FORMAT_RGB_565) {
+    // Only 5 bits of information per channel except G which has 6.
+    epsilon = 1.0f / (1 << 5);
+  } else if (rgb.isFloat) {
+    epsilon = 0.0005f;  // Half precision floats are not that precise.
+  }
+
+  float pixel_read[4];
+  for (uint32_t j = 0; j < rgb.height; ++j) {
+    for (uint32_t i = 0; i < rgb.width; ++i) {
+      // Generate some arbitrary pixel values.
+      const float pixel_to_write[4] = {
+          0.0f + static_cast<float>(i) / rgb.width,
+          0.5f + static_cast<float>(j) / (rgb.height * 2),
+          1.0f - static_cast<float>(i + j) / ((rgb.width + rgb.height) * 2),
+          1.0f - static_cast<float>(i) / rgb.width};
+
+      avifSetRGBAPixel(&rgb, i, j, &color_space, pixel_to_write);
+      avifGetRGBAPixel(&rgb, i, j, &color_space, pixel_read);
+      EXPECT_NEAR(pixel_read[0], pixel_to_write[0], epsilon);
+      EXPECT_NEAR(pixel_read[1], pixel_to_write[1], epsilon);
+      EXPECT_NEAR(pixel_read[2], pixel_to_write[2], epsilon);
+      if (avifRGBFormatHasAlpha(rgb_format)) {
+        EXPECT_NEAR(pixel_read[3], pixel_to_write[3], epsilon);
+      } else {
+        EXPECT_EQ(pixel_read[3], 1.0f);
+      }
```

Comment:
Note that the `avifImage->float->avifImage` round-trip would allow for exact value equality comparison instead of `epsilon` equivalence. It can be added later on if useful (the current test does not guarantee that there is no color drift when converting back-and-forth multiple times).

Reply:
I'm sorry I don't really understand what test you have in mind when you talk about `avifImage->float->avifImage`.

Reply:
My bad, I meant `avifRGBImage->float->avifRGBImage`. Right now this is testing `float->avifRGBImage->float` which has some loss; the other way around should be lossless and easier to test. But I guess this test exists in this way for a reason, so please disregard this note if irrelevant.

Reply:
Ha thanks I understand now. The thing is, how would get an avifRGBImage in all these different RGB formats to run the test?
The simplest way is to use avifSetRGBAPixel, which essentially leads to a test like this.
I guess you could take yuv image and convert it to rgb, but again I think that muddies the waters because the yuv->rgb conversion could be using these two functions at some point.

Actually I just tried adding a test like this: `avifImage with FillImageGradient->avifRGBImage->float->avifRGBImage->avifImage` and then use GetPSNR to compare at the end. Aaaand it doesn't work because of clipping issues. FillImageGradient create YUV values that you can't get by converting from RGB. E.g. YUV(0,0,0) converts to RGB(0,135,0) but if you convert that back you get YUV(79,83,71) (for BT.601)
You can double check e.g. on https://www.mikekohn.net/file_formats/yuv_rgb_converter.php by clicking the "yuv2rgb" button then the "rgb2yuv" button. So it would need to be changed to create valid values (maybe by converting from RGB? inception...).

Anyway you did say I could disregard this and I have a headache now so I will say this test is good enough for now :)

Reply:
> how would get an avifRGBImage in all these different RGB formats to run the test?

See a similar test at https://chromium.googlesource.com/codecs/libwebp2/+/49cb845ecae4c3856e7093eddd305b2a7c63b063/tests/test_color_precision.cc#74  
It should be adapted to each format of course (trivial for ARGB permutations, special cases for float and rgb565).

> Actually I just tried adding a test like this: `avifImage with FillImageGradient->avifRGBImage->float->avifRGBImage->avifImage` and then use GetPSNR to compare at the end. Aaaand it doesn't work because of clipping issues.

This approach could work if the comparison was made on `avifRGBImage` and not `avifImage`, but as you said it may not be bijective so some input ARGB combinations would not be tested. But as a quick additional test it could be enough. Let me try that.

> Anyway you did say I could disregard this and I have a headache now so I will say this test is good enough for now :)

Of course. This is just further thorough testing, the current one is already great. Thanks for adding tests.